### PR TITLE
Added video support to loading manager

### PR DESCRIPTION
**Pull request by osiriswrecks**
We needed a basic way to track loading progress for video textures on a project at work, so I added VideoLoader and VideoTextureLoader objects based on existing code.

This is my first pull request...well, ever, so I'm happy to clean up and reduce duplicated code (and right now, there's a good chunk of that because I copied the ImageLoader and TextureLoader files and tweaked things to work for video). I'd love to see this functionality added to the mainstream build if others find it useful.

### CODE REVIEW

File: src/loaders/VideoLoader.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+				scope.manager.itemEnd( url );
+
+			}, 0 );
+
+			return cached;
+
+		}
+
+		var image = document.createElementNS( 'http://www.w3.org/1999/xhtml', 'video' );
```
Comment (Mugen87) [inline src/loaders/VideoLoader.js]: `image`? Maybe `video` would be a better name. I guess this is a copy/paste error from `ImageLoader`, right?

File: src/loaders/VideoLoader.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+			image.removeEventListener( 'error', onImageError, false );
+
+			if ( onError ) onError( event );
+
+			scope.manager.itemEnd( url );
+			scope.manager.itemError( url );
+
+		}
+
+		image.addEventListener( 'loadeddata', onImageLoad, false );
```
Comment (Mugen87) [inline src/loaders/VideoLoader.js]: Why `loadeddata` and not `canplaythrough` or `canplay`?

Comment (WestLangley) [reply src/loaders/VideoLoader.js]: How about 'load'  ?

Comment (WestLangley) [reply src/loaders/VideoLoader.js]: 'loaded' seems reasonable. The event is what did happen, not what can happen.

**Comment by osiriswrecks** [comment]
I’m happy to make the change. I think I chose ‘loadeddata’ because ‘loaded’ wasn’t working for me and it wasn’t listed in the event call list that fires for video (can find a source when I’m not on my phone).

**Comment by Mugen87** [comment]
Yeah, the video element is something special since we don't want to wait until the entire resource is downloaded. Since it's a stream, it's necessary to work with different events like `loadeddata`, `canplay` or `canplaythrough`. The problem is that it seems inconsistent if these events trigger an `onLoad()` callback. `onLoad()` in three.js is always used when the given resource is entirely loaded. So maybe we need a different nomenclature for such a loader.

**Comment by osiriswrecks** [comment]
@Mugen87 I didn't consider it. If you think `VideoLoader` isn't necessary on it's own I'll merge the two together.

**Comment by Mugen87** [comment]
Let's see what the others think about the idea 😊

**Comment by Mugen87** [comment]
Instead of creating a `VideoLoader`, we should consider to implement something like a `MediaLoader` which does only rely on the interface of [HTMLMediaElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement). In this way, we could also use it for audio streams. If you currently want to use a background or ambient music in your app, there is now way to manage the loading state with `three.js`. `AudioLoader` loads the entire audio and decodes it. Hence, it is thus not suitable for loading bigger files like music which you want to stream.

**Comment by yomotsu** [comment]
We could load a media file as a binary data via xhr, then make a video or audio element as a Blob or data URL
Then, we can detect loaded-entirely event

How do you think?

**Comment by Mugen87** [comment]
As I've mentioned before, I think it's better to stream audio and video data than loading the entire file. In context of audio data, I refer to files for background music and not single sound effects (like steps, shots or UI-feedback).

We could try to identify the media over its file extension and then create the respective element.

**Comment by osiriswrecks** [comment]
@Mugen87 I'm not sure we should be splitting audio like that, even if it makes sense technically. Different users will classify their audio differently, and may end up using an AudioLoader when they're supposed to be using a MediaLoader and vise versa.

To @yomotsu 's comment, have we considered using Media Source Extensions to load video content? It would make delivery more efficient and from what I can tell helps avoid issue #8110 on older iOS/Mac devices.

**Comment by mrdoob** [comment]
I agree with @Mugen87.
Can you merge both loaders into a single `VideoTextureLoader` for now?
We can always create a `VideoLoader` in the future if there is a use case.

**Comment by osiriswrecks** [comment]
@mrdoob Sure can. Do we have any preferences on whether it should just be video, or should this be a more abstract media loader with MSE support so we can handle audio and video?

@Mugen87, I know I was initially against the idea of merging them together, but the more I've worked with MSE on my own I've realized it makes a lot of sense to combine it all together as a single streaming object. What I'm saying is, you were totally right :)

**Comment by osiriswrecks** [comment]
I finally got around to merging these two together. Here's a working example on my own server, which I can also commit if you'd like: https://dev.lunadigital.tv/three/examples/webgl_loader_texture_video.html

**Comment by osiriswrecks** [comment]
I noticed my latest push didn't pass the CI check because of a trailing space in the code. I can't figure out why for the life of me - there's no obvious trailing space at the line it references. Anything obvious I'm missing?

**Comment by Mugen87** [comment]
After thinking about this PR once more I doubt that `VideoTextureLoader` would be much useful:

- Compared to images, certain formats like OGG are not supported by all browsers. Defining multiple sources via HTML seems more flexible to me than using fixed URLs with `VideoTextureLoader`. That would require a capabilities check before you load the video from the backend.
- Attributes and config settings of video elements might vary from use case to use case e.g. (muted, looped etc.). Images are more simple in this context.
- Projects might prefer events like `canplay` or `canplaythrough` over `loadedmetadata`.

Hence, I guess it's better to let the application provide the (configured) video element to `VideoTexture`.

**Pull request closed without merging**